vicodyn/src/vicodyn/proxy.cpp

```diff
@@ -371,24 +442,23 @@ class vicodyn_dispatch_t : public std::enable_shared_from_this<vicodyn_dispatch_
         return std::shared_ptr<dispatch<app_tag>>(shared_from_this(), &forward_dispatch);
     }
 
-    auto choose_endpoint() -> void {
-        endpoint.peer = proxy.balancer->choose_peer(request_context, buffer.headers, buffer.event);
+    auto choose_endpoint(const hpack::headers_t& headers, const std::string& event) -> void {
+        endpoint.peer = proxy.balancer->choose_peer(request_context, headers, event);
         request_context->mark_used_peer(endpoint.peer);
         endpoint.logger = std::make_unique<blackhole::wrapper_t>(*proxy.logger,
                         blackhole::attributes_t{{"peer", endpoint.peer->uuid()}});
         endpoint.start_time = clock_t::now();
-        auto u = endpoint.peer->open_stream<io::node::enqueue>(shared_backward_dispatch(), buffer.headers,
-                        proxy.app_name, buffer.event);
+
+        auto u = endpoint.peer->open_stream<io::node::enqueue>(shared_backward_dispatch(), headers,
+                        proxy.app_name, event);
         endpoint.forward_stream = std::move(safe_stream_t(std::move(u)));
     }
 
-    auto enqueue_unsafe(const hpack::headers_t& headers, const std::string& event)
-                -> std::shared_ptr<dispatch<app_tag>> {
-        buffer.event = std::move(event);
-        buffer.headers = std::move(headers);
-
+    auto enqueue_unsafe(const hpack::headers_t& headers, const std::string& event) -> std::shared_ptr<dispatch<app_tag>> {
         try {
-            choose_endpoint();
+            buffer.set_event(headers, event);
```

Comment:
Event is logically a part of buffer at this point, so you can move event parts into buffer and call choose_endpoint on `buffer_t` argument.